### PR TITLE
2.1: Update repo to allow partial builds

### PR DESCRIPTION
**Pull request by natemcmaster**
Part of unifying to fewer repos means we need a mechanism for building only _some_ packages from a repo in servicing builds. In the past, we have treated repositories as a unit of work (if one thing patches in a repo, we re-release the whole thing.) With fewer repos, that strategy doesn't work as well. This PR adds targets and tools to support only building some packages from the repo.

Changes:
* Add a tool which generates a "baseline" file. This helps ensure packages that are not patching in a servicing update maintain the same version and dependencies.
* Add targets to resolve `<Reference>` items against this baseline
* Add a new file `eng/PatchConfig.props` in which dev should add the ID of the package being serviced in a particular release.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -18,6 +17,20 @@
 
     <ArtifactsDir>$(MSBuildThisFileDirectory)artifacts\</ArtifactsDir>
     <PackageOutputPath>$(MSBuildThisFileDirectory)artifacts\build\</PackageOutputPath>
+
+    <IsBenchmarkProject Condition="$(MSBuildProjectName.EndsWith('.Performance'))">true</IsBenchmarkProject>
+    <IsTestProject Condition="$(MSBuildProjectName.EndsWith('Tests')) OR $(MSBuildProjectName.EndsWith('.Test'))">true</IsTestProject>
```
Comment (Eilon) [inline Directory.Build.props]: I have slightly mixed feelings about "convention over configuration" for these types of things, but I'm fine starting out here and seeing how it goes.

File: eng/Dependencies.props [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    <LatestPackageReference Include="BenchmarkDotNet" Version="0.10.13" />
+    <LatestPackageReference Include="Microsoft.NET.Test.Sdk" Version="15.9.0" />
+    <LatestPackageReference Include="Moq" Version="4.10.0" />
+    <LatestPackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <LatestPackageReference Include="xunit.analyzers" Version="0.10.0" />
+    <LatestPackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
+    <LatestPackageReference Include="xunit" Version="2.4.0" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(IsTestProject)' == 'true' ">
```
Comment (Eilon) [inline eng/Dependencies.props]: Does "refactoring" dependencies work well in VS? I guess we don't support changing dependencies in VS anyway and you always have to hand-edit projects?

File: eng/targets/Packaging.targets [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  <PropertyGroup>
+    <PackageId Condition=" '$(PackageId)' == '' ">$(AssemblyName)</PackageId>
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(IsPackable)' != 'false' AND '$(PatchVersion)' != '0' ">
+    <IsPackable>$(PackagesInPatch.Contains(' $(PackageId);'))</IsPackable>
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(IsPackable)' == 'false' AND '$(BaselinePackageVersion)' != '' ">
+    <Version>$(BaselinePackageVersion).0</Version>
```
Comment (Eilon) [inline eng/targets/Packaging.targets]: Not sure I get what this section is doing. Maybe add a comment?

File: shared/Microsoft.Extensions.CommandLineUtils.Sources/Microsoft.Extensions.CommandLineUtils.Sources.projitems [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Eilon) [inline shared/Microsoft.Extensions.CommandLineUtils.Sources/Microsoft.Extensions.CommandLineUtils.Sources.projitems]: Question: What is a `.projitems` file meant to represent? Why even more factoring?

File: eng/Dependencies.props (hunk 2 above)
Comment (natemcmaster) [reply eng/Dependencies.props]: Hand-edit. Our build is custom enough that we don't attempt to support using VS to manipulate references.

File: eng/targets/Packaging.targets (hunk 3 above)
Comment (natemcmaster) [reply eng/targets/Packaging.targets]: Added.

File: shared/Microsoft.Extensions.CommandLineUtils.Sources/Microsoft.Extensions.CommandLineUtils.Sources.projitems (hunk 4 above)
Comment (natemcmaster) [reply shared/Microsoft.Extensions.CommandLineUtils.Sources/Microsoft.Extensions.CommandLineUtils.Sources.projitems]: This is using the "Shared Project" feature in VS (which works as far back as VS 2013). It's just UI sugar for working with files that are compiled into multiple projects.

**Pull request merged**